transport/src/main/java/io/netty/channel/pool/FixedChannelPool.java

```diff
@@ -441,19 +442,35 @@ public void acquired() {
 
     @Override
     public void close() {
+        try {
+            closeAsync().await();
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Closes the pool in an async manner.
+     *
+     * @return Future which represents completion of the close task
+     */
+    public Future<Void> closeAsync() {
         if (executor.inEventLoop()) {
-            close0();
+            return close0();
         } else {
-            executor.submit(new Runnable() {
+            return executor.submit(new Callable<Future<Void>>() {
```

Comment:
This is a little surprising.   Why not:

1.  Make a Promise
2.  use executor.execute() to run a Runnable to call close()
3.  Add a listener to the close future, to update the promise.

That would avoid the need to call wait uninterruptibly, and match the existing patterns in AbstractChannelHandlerContext.

Reply:
Thanks for the suggestion. This is much cleaner. Code changed.